lib/elixir/test/elixir/fixtures/dialyzer/boolean_check.ex

```diff
@@ -0,0 +1,17 @@
+defmodule Dialyzer.BooleanCheck do
+  def and_check(arg) do
+    arg and arg
+  end
+
+  def and_check_optimized(arg) do
+    is_integer(arg) and arg > 10
+  end
+
+  def or_check(arg) do
+    arg or arg
```

Comment:
Should we check more interesting cases, where dialyzer can infer that a pattern can never match? If dialyzers warns there is it valid? Previously it would not warn (I believe).
e.g.
```elixir
def bad_or_check(arg) when is_integer(arg) do
  arg < :infinity or arg
end
```